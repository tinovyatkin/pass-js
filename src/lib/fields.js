'use strict';

/**
 * Field accessors class
 */

const only = require('./only');
const { getW3CDateString } = require('./w3cdate');

class Fields {
  /**
   * Creates an instance of Fields
   *
   * @param {import('../pass')} pass - parent Pass class
   * @param {string} key - key name that these fields are bound to
   * @memberof Fields
   */
  constructor(pass, key) {
    this.pass = pass;
    this.key = key;
    Object.preventExtensions(this);
  }

  /**
   * Returns an array of all fields.
   *
   * @returns {{key: string, value: string, label: string}[]}
   * @memberof Fields
   */
  all() {
    if (!(this.key in this.pass.structure)) this.pass.structure[this.key] = [];
    return this.pass.structure[this.key];
  }

  /**
   * Adds a field to the end of the list
   *
   * @param {string | {key: string, label: string, value: string}} key - Field key or object with all fields
   * @param {string} [label] - Field label (optional)
   * @param {string} [value] - Field value
   * @param {Object} [options] - Other field options (e.g. dateStyle)
   * @returns {Fields}
   * @memberof Fields
   */
  add(key, label, value, options) {
    if (arguments.length > 1) {
      this.remove(key);
      const field = { key, value };
      if (label) field.label = label;
      if (options) Object.assign(field, options);
      this.all().push(field);
    } else if (Array.isArray(key)) {
      for (const k of key) this.add(k);
    } else {
      this.remove(key.key);
      // save object copy
      this.all().push(Object.assign({}, key));
    }
    return this;
  }

  /**
   * Returns a field
   *
   * @param {string} key
   * @returns {{key: string, label: string, value: string}} If field exists, returns an object with common keys and rest of keys
   * @memberof Fields
   */
  get(key) {
    return this.all().find(v => v.key === key);
  }

  /**
   * Sets value field for a given key
   *
   * @param {string} key
   * @param {string} value
   * @memberof Fields
   */
  setValue(key, value) {
    if (typeof key !== 'string')
      throw new TypeError(
        `key for setValue must be a string, received ${typeof key}`,
      );
    if (typeof value !== 'string')
      throw new TypeError(
        `value for setValue must be a string, received ${typeof value}`,
      );
    const field = this.get(key);
    if (!field) return this.add({ key, value });
    field.value = value;
    return this;
  }

  /**
   * Removes a given field.
   *
   * @param {string} key
   * @returns {Fields}
   * @memberof Fields
   */
  remove(key) {
    const idx = this.all().findIndex(v => v.key === key);
    if (idx > -1) {
      this.all().splice(idx, 1);
      // remove property completely if there is no fields left
      if (this.all().length === 0) this.clear();
    }
    return this;
  }

  /**
   * Set a field as Date value with appropriated options
   *
   * @param {string} key
   * @param {string} label
   * @param {Date} date
   * @param {{dateStyle?: string, ignoresTimeZone?: boolean, isRelative?: boolean, timeStyle?:string}} [formatOptions]
   * @returns {Fields}
   * @throws if date is not a Date or invalid Date
   * @memberof Fields
   */
  setDateTime(key, label, date, formatOptions = {}) {
    if (!(date instanceof Date) || !isFinite(date))
      throw new Error(
        'First parameter of setDateTime must be an instance of Date',
      );
    //  Either specify both a date style and a time style, or neither.
    if (
      (!('dateStyle' in formatOptions) && 'timeStyle' in formatOptions) ||
      ('dateStyle' in formatOptions && !('timeStyle' in formatOptions))
    )
      throw new Error(
        'Either specify both a date style and a time style, or neither',
      );
    // adding
    this.add(
      Object.assign(
        {
          key,
          label,
          value: getW3CDateString(date),
        },
        only(
          formatOptions,
          'dateStyle timeStyle ignoresTimeZone isRelative changeMessage',
        ),
      ),
    );
    return this;
  }

  /**
   * Removes all fields.
   *
   * @returns {Fields}
   * @memberof Fields
   */
  clear() {
    delete this.pass.structure[this.key];
    return this;
  }
}

module.exports = Fields;
