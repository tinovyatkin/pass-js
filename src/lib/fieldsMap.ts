'use strict';

import { Field, FieldDescriptor, DataStyleFormat } from '../interfaces';

import { getW3CDateString } from './w3cdate';

export class FieldsMap extends Map<string, FieldDescriptor> {
  /**
   * Returns Map as array of objects with key moved inside like a key property
   */
  toJSON(): Field[] | undefined {
    if (!this.size) return undefined;
    return [...this].map(
      ([key, data]): Field => {
        // Remap Date objects to string
        if (data.value instanceof Date)
          data.value = getW3CDateString(data.value);
        return { key, ...data };
      },
    );
  }

  /**
   * Adds a field to the end of the list
   *
   * @param {Field} field - Field key or object with all fields
   * @returns {FieldsMap}
   * @memberof FieldsMap
   */
  add(field: Field): FieldsMap {
    const { key, ...data } = field;
    if (typeof key !== 'string')
      throw new TypeError(
        `To add a field you must provide string key value, received ${typeof key}`,
      );
    if (!('value' in data))
      throw new TypeError(
        `To add a field you must provide a value field, received: ${JSON.stringify(
          data,
        )}`,
      );
    this.set(key, data);
    return this;
  }

  /**
   * Sets value field for a given key, without changing the rest of field properties
   *
   * @param {string} key
   * @param {string} value
   * @memberof FieldsMap
   */
  setValue(key: string, value: string): FieldsMap {
    if (typeof key !== 'string')
      throw new TypeError(
        `key for setValue must be a string, received ${typeof key}`,
      );
    if (typeof value !== 'string')
      throw new TypeError(
        `value for setValue must be a string, received ${typeof value}`,
      );
    const field = this.get(key) || { value };
    field.value = value;
    this.set(key, field);
    return this;
  }

  /**
   * Set a field as Date value with appropriated options
   *
   * @param {string} key
   * @param {string} label
   * @param {Date} date
   * @param {{dateStyle?: string, ignoresTimeZone?: boolean, isRelative?: boolean, timeStyle?:string}} [formatOptions]
   * @returns {FieldsMap}
   * @throws if date is not a Date or invalid Date
   * @memberof FieldsMap
   */
  setDateTime(
    key: string,
    label: string,
    date: Date,
    {
      dateStyle,
      ignoresTimeZone,
      isRelative,
      timeStyle,
    }: {
      dateStyle?: DataStyleFormat;
      ignoresTimeZone?: boolean;
      isRelative?: boolean;
      timeStyle?: DataStyleFormat;
    } = {},
  ): FieldsMap {
    if (typeof key !== 'string')
      throw new TypeError(`Key must be a string, received ${typeof key}`);
    if (typeof label !== 'string')
      throw new TypeError(`Label must be a string, received ${typeof label}`);
    if (!(date instanceof Date))
      throw new TypeError(
        'Third parameter of setDateTime must be an instance of Date',
      );
    //  Either specify both a date style and a time style, or neither.
    if (!!dateStyle !== !!timeStyle)
      throw new ReferenceError(
        'Either specify both a date style and a time style, or neither',
      );
    // adding
    this.set(key, {
      label,
      value: date,
      dateStyle,
      ignoresTimeZone,
      isRelative,
      timeStyle,
    });

    return this;
  }
}
