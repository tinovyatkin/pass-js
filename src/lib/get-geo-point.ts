/**
 * Returns normalized geo point object from geoJSON, {lat, lng} or {lattitude,longutude,altitude}
 *
 * @param {number[] | { lat: number, lng: number, alt?: number } | { longitude: number, latitude: number, altitude?: number }} point
 * @returns {{ longitude: number, latitude: number, altitude?: number }}
 * @throws on unknown point format
 */
export function getGeoPoint(
  point:
    | number[]
    | { lat: number; lng: number; alt?: number }
    | { longitude: number; latitude: number } & (
        | { altitude?: number }
        | { elevation?: number }),
): { longitude: number; latitude: number; altitude?: number } {
  if (!point) throw new Error("Can't get coordinates from undefined");

  // GeoJSON Array [longitude, latitude(, elevation)]
  if (Array.isArray(point)) {
    if (point.length < 2 || !point.every(n => Number.isFinite(n)))
      throw new Error(
        `Invalid GeoJSON array of numbers, length must be 2 to 3, received ${
          point.length
        }`,
      );
    return {
      longitude: point[0],
      latitude: point[1],
      altitude: point[2],
    };
  }

  // it can be an object with both lat and lng properties
  if ('lat' in point && 'lng' in point) {
    return {
      longitude: point.lng,
      latitude: point.lat,
      altitude: point.alt,
    };
  }

  if ('longitude' in point && 'latitude' in point) {
    // returning a copy
    return {
      longitude: point.longitude,
      latitude: point.latitude,
      altitude:
        'altitude' in point
          ? point.altitude
          : 'elevation' in point
          ? point.elevation
          : undefined,
    };
  }

  // If we are here it means we can't understand what a hell is it
  throw new Error(`Unknown geo point format: ${JSON.stringify(point)}`);
}
