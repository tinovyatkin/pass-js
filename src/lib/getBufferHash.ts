'use strict';

import { createHash } from 'crypto';

/**
 *
 * @param {Buffer} buffer
 * @returns {string}
 */
export function getBufferHash(buffer: Buffer | string): string {
  // creating hash
  const sha = createHash('sha1');
  sha.update(buffer);
  return sha.digest('hex');
}
