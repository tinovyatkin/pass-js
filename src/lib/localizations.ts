/**
 * Class to handle Apple pass localizations
 * @see {@link https://developer.apple.com/library/archive/documentation/UserExperience/Conceptual/PassKit_PG/Creating.html#//apple_ref/doc/uid/TP40012195-CH4-SW54}
 *
 */

import { createReadStream } from 'fs';
import { createInterface } from 'readline';
import * as path from 'path';

import * as glob from 'fast-glob';

/**
 * @see {@link https://github.com/justinklemm/i18n-strings-files/blob/dae303ed60d9d43dbe1a39bb66847be8a0d62c11/index.coffee#L100}
 * @param {string} filename - path to pass.strings file
 */
export async function readLprojStrings(
  filename: string,
): Promise<Map<string, string>> {
  const res = new Map() as Map<string, string>;
  let nextLineIsComment = false;
  const rl = createInterface(
    createReadStream(filename, { encoding: 'utf16le' }),
  );
  for await (const line of rl) {
    // skip empty lines
    const l = line.trim();
    if (!l) continue;
    // check if starts with '/*' and skip comments
    if (nextLineIsComment || l.startsWith('/*')) {
      nextLineIsComment = !l.endsWith('*/');
      continue;
    }
    // check for first quote, assignment operator, and final semicolon
    const test = /^"(?<msgId>.+)"\s*=\s*"(?<msgStr>.+)"\s*;/.exec(l);
    if (!test) continue;
    let { msgId, msgStr } = test.groups as { msgId: string; msgStr: string };
    //  convert escaped quotes
    msgId = msgId.replace(/\\"/g, '"');
    msgStr = msgStr.replace(/\\"/g, '"');
    //  convert escaped new lines
    msgId = msgId.replace(/\\n/g, '\n');
    msgStr = msgStr.replace(/\\n/g, '\n');
    res.set(msgId, msgStr);
  }
  return res;
}

/**
 * Converts given translations map into UTF-16 encoded buffer in .lproj format
 *
 * @param {Map.<string, string>} strings
 */
export function getLprojBuffer(strings: Map<string, string>): Buffer {
  return Buffer.from(
    [...strings]
      .map(
        ([key, value]) =>
          `"${key
            .replace(/\n/gm, '\\n') // escape new lines
            .replace(/"/g, '"')}" = "${value
            .replace(/\n/gm, '\\n') // escape new lines
            .replace(/"/g, '"')}";`,
      )
      .join('\n'),
    'utf16le',
  );
}

/**
 * Localizations class Map<lang, Map<key, translation>>
 */
export class Localizations extends Map<string, Map<string, string>> {
  add(lang: string, values: { [k: string]: string }): void {
    const map: Map<string, string> = this.get(lang) || new Map();
    for (const [key, value] of Object.entries(values)) {
      map.set(key, value);
    }
    if (!this.has(lang)) this.set(lang, map);
  }

  toArray(): { path: string; data: Buffer }[] {
    return [...this].map(([lang, map]) => ({
      path: `${lang}.lproj/pass.strings`,
      data: getLprojBuffer(map),
    }));
  }

  /**
   * Loads available localizations from given folder path
   *
   * @param {string} dirPath
   */
  async load(dirPath: string): Promise<void> {
    for await (const file of glob.stream(
      [path.join(dirPath, '*.lproj/pass.strings')],
      {
        onlyFiles: true,
        deep: 1,
      },
    )) {
      if (typeof file !== 'string') continue;
      /*
          The file is always ends with something like /zh_CN.lproj/pass.string
          So, taking the end will give us a language
          */
      const language = path
        .basename(path.dirname(file), '.lproj')
        .replace(/_/g, '-'); // we will tolerate zh_CN as well as zh-CN
      /**
       * @see {@link https://stackoverflow.com/questions/8758340/regex-to-detect-locales}
       */
      if (
        !/^[A-Za-z]{2,4}(-([A-Za-z]{4}|\d{3}))?(-([A-Za-z]{2}|\d{3}))?$/.test(
          language,
        )
      ) {
        console.warn(
          'Localization file %s was not loaded because "%s" is not looks like a valid language/locale code',
          file,
          language,
        );
        return;
      }
      this.set(language, await readLprojStrings(file));
    }
  }
}
