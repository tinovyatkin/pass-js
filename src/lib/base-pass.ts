import { ApplePass } from '../interfaces';
import { BARCODES_FORMAT, STRUCTURE_FIELDS } from '../constants';

import { PassColor } from './pass-color';
import { PassImages } from './images';
import { Localizations } from './localizations';
import { getGeoPoint } from './get-geo-point';
import { PassStructure } from './pass-structure';

const STRUCTURE_FIELDS_SET = new Set([...STRUCTURE_FIELDS, 'nfc']);

export class PassBase extends PassStructure {
  images: PassImages;
  protected localization = new Localizations();

  constructor(fields: Partial<ApplePass> = {}, images?: PassImages) {
    super(fields);

    // restore via setters
    for (const [key, value] of Object.entries(fields)) {
      if (!STRUCTURE_FIELDS_SET.has(key) && key in this) {
        this[key] = value;
      }
    }

    // copy images
    this.images = new PassImages(images);
  }

  // Returns the pass.json object (not a string).
  toJSON(): Partial<ApplePass> {
    return { ...this.fields, formatVersion: 1 };
  }

  get passTypeIdentifier(): string | undefined {
    return this.fields.passTypeIdentifier;
  }

  set passTypeIdentifier(v: string | undefined) {
    if (!v) delete this.fields.passTypeIdentifier;
    else this.fields.passTypeIdentifier = v;
  }

  get teamIdentifier(): string | undefined {
    return this.fields.teamIdentifier;
  }

  set teamIdentifier(v: string | undefined) {
    if (!v) delete this.fields.teamIdentifier;
    else this.fields.teamIdentifier = v;
  }

  get serialNumber(): string | undefined {
    return this.fields.serialNumber;
  }

  set serialNumber(v: string | undefined) {
    if (!v) delete this.fields.serialNumber;
    else this.fields.serialNumber = v;
  }

  /**
   *  Indicates that the pass is void—for example, a one time use coupon that has been redeemed.
   *
   * @type {boolean}
   */
  get voided(): boolean {
    return !!this.fields.voided;
  }

  set voided(v: boolean) {
    if (v) this.fields.voided = true;
    else delete this.fields.voided;
  }

  /**
   * Date and time when the pass expires.
   *
   */
  get expirationDate(): ApplePass['expirationDate'] {
    if (typeof this.fields.expirationDate === 'string')
      return new Date(this.fields.expirationDate);
    return this.fields.expirationDate;
  }
  set expirationDate(v: ApplePass['expirationDate']) {
    if (!v) delete this.fields.expirationDate;
    else {
      if (typeof v === 'string' || v instanceof Date)
        this.fields.expirationDate = v;
      else
        throw new TypeError(
          `expirationDate should be either string or Date, received ${typeof v}`,
        );
    }
  }

  /**
   * Date and time when the pass becomes relevant. For example, the start time of a movie.
   * Recommended for event tickets and boarding passes; otherwise optional.
   *
   * @type {string | Date}
   */
  get relevantDate(): ApplePass['relevantDate'] {
    if (typeof this.fields.relevantDate === 'string')
      return new Date(this.fields.relevantDate);
    return this.fields.relevantDate;
  }

  set relevantDate(v: ApplePass['relevantDate']) {
    if (!v) delete this.fields.relevantDate;
    else {
      if (typeof v === 'string' || v instanceof Date)
        this.fields.relevantDate = v;
      else
        throw new TypeError(
          `relevantDate should be either string or Date, received ${typeof v}`,
        );
    }
  }

  /**
   * A list of iTunes Store item identifiers for the associated apps.
   * Only one item in the list is used—the first item identifier for an app
   * compatible with the current device.
   * If the app is not installed, the link opens the App Store and shows the app.
   * If the app is already installed, the link launches the app.
   */
  get associatedStoreIdentifiers(): ApplePass['associatedStoreIdentifiers'] {
    return this.fields.associatedStoreIdentifiers;
  }
  set associatedStoreIdentifiers(v: ApplePass['associatedStoreIdentifiers']) {
    if (!v) {
      delete this.fields.associatedStoreIdentifiers;
      return;
    }
    const arrayOfNumbers = v.filter(n => Number.isInteger(n));
    if (arrayOfNumbers.length > 0)
      this.fields.associatedStoreIdentifiers = arrayOfNumbers;
    else delete this.fields.associatedStoreIdentifiers;
  }

  /**
   * Brief description of the pass, used by the iOS accessibility technologies.
   * Don’t try to include all of the data on the pass in its description,
   * just include enough detail to distinguish passes of the same type.
   */
  get description(): string | undefined {
    return this.fields.description;
  }

  set description(v: string | undefined) {
    if (!v) delete this.fields.description;
    else this.fields.description = v;
  }

  /**
   * Display name of the organization that originated and signed the pass.
   */
  get organizationName(): string | undefined {
    return this.fields.organizationName;
  }
  set organizationName(v: string | undefined) {
    if (!v) delete this.fields.organizationName;
    else this.fields.organizationName = v;
  }

  /**
   * Optional for event tickets and boarding passes; otherwise not allowed.
   * Identifier used to group related passes.
   * If a grouping identifier is specified, passes with the same style,
   * pass type identifier, and grouping identifier are displayed as a group.
   * Otherwise, passes are grouped automatically.
   * Use this to group passes that are tightly related,
   * such as the boarding passes for different connections of the same trip.
   */
  get groupingIdentifier(): string | undefined {
    return this.fields.groupingIdentifier;
  }

  set groupingIdentifier(v: string | undefined) {
    if (!v) delete this.fields.groupingIdentifier;
    else this.fields.groupingIdentifier = v;
  }

  /**
   * If true, the strip image is displayed without a shine effect.
   * The default value prior to iOS 7.0 is false.
   * In iOS 7.0, a shine effect is never applied, and this key is deprecated.
   */
  get suppressStripShine(): boolean {
    return !!this.fields.suppressStripShine;
  }
  set suppressStripShine(v: boolean) {
    if (!v) delete this.fields.suppressStripShine;
    else this.fields.suppressStripShine = true;
  }

  /**
   * Text displayed next to the logo on the pass.
   */
  get logoText(): string | undefined {
    return this.fields.logoText;
  }
  set logoText(v: string | undefined) {
    if (!v) delete this.fields.logoText;
    else this.fields.logoText = v;
  }

  /**
   * The URL of a web service that conforms to the API described in PassKit Web Service Reference.
   * The web service must use the HTTPS protocol; the leading https:// is included in the value of this key.
   * On devices configured for development, there is UI in Settings to allow HTTP web services.
   *
   * @see {@link https://developer.apple.com/library/archive/documentation/PassKit/Reference/PassKit_WebService/WebService.html#//apple_ref/doc/uid/TP40011988}
   */
  get webServiceURL(): URL | string | undefined {
    return this.fields.webServiceURL;
  }
  set webServiceURL(v: URL | string | undefined) {
    if (!v) {
      delete this.fields.webServiceURL;
      return;
    }

    // validating URL, it will throw on bad value
    const url = v instanceof URL ? v : new URL(v);
    if (url.protocol !== 'https:')
      throw new TypeError(`webServiceURL must be on HTTPS!`);
    this.fields.webServiceURL = url;
  }

  /**
   * The authentication token to use with the web service.
   * The token must be 16 characters or longer.
   */
  get authenticationToken(): string | undefined {
    return this.fields.authenticationToken;
  }
  set authenticationToken(v: string | undefined) {
    if (!v) {
      delete this.fields.authenticationToken;
      return;
    }

    if (typeof v !== 'string')
      throw new TypeError(
        `authenticationToken must be a string, received ${typeof v}`,
      );
    if (v.length < 16)
      throw new TypeError(
        `authenticationToken must must be 16 characters or longer`,
      );
    this.fields.authenticationToken = v;
  }

  /**
   * Background color of the pass, specified as an CSS-style RGB triple.
   *
   * @example rgb(23, 187, 82)
   */
  get backgroundColor():
    | [number, number, number]
    | string
    | undefined
    | PassColor {
    if (!(this.fields.backgroundColor instanceof PassColor)) return undefined;
    return this.fields.backgroundColor.getRGB();
  }
  set backgroundColor(
    v: string | [number, number, number] | undefined | PassColor,
  ) {
    if (!v) {
      delete this.fields.backgroundColor;
      return;
    }
    if (!(this.fields.backgroundColor instanceof PassColor))
      this.fields.backgroundColor = new PassColor(v);
    else this.fields.backgroundColor.set(v);
  }

  /**
   * Foreground color of the pass, specified as a CSS-style RGB triple.
   *
   * @example rgb(100, 10, 110)
   */
  get foregroundColor():
    | [number, number, number]
    | string
    | undefined
    | PassColor {
    if (!(this.fields.foregroundColor instanceof PassColor)) return undefined;
    return this.fields.foregroundColor.getRGB();
  }
  set foregroundColor(
    v: string | [number, number, number] | PassColor | undefined,
  ) {
    if (!v) {
      delete this.fields.foregroundColor;
      return;
    }
    if (!(this.fields.foregroundColor instanceof PassColor))
      this.fields.foregroundColor = new PassColor(v);
    else this.fields.foregroundColor.set(v);
  }

  /**
   * Color of the label text, specified as a CSS-style RGB triple.
   *
   * @example rgb(255, 255, 255)
   */
  get labelColor(): [number, number, number] | string | undefined | PassColor {
    if (!(this.fields.labelColor instanceof PassColor)) return undefined;
    return this.fields.labelColor.getRGB();
  }
  set labelColor(v: string | [number, number, number] | PassColor | undefined) {
    if (!v) {
      delete this.fields.labelColor;
      return;
    }
    if (!(this.fields.labelColor instanceof PassColor))
      this.fields.labelColor = new PassColor(v);
    else this.fields.labelColor.set(v);
  }

  /**
   * Maximum distance in meters from a relevant latitude and longitude that the pass is relevant.
   * This number is compared to the pass’s default distance and the smaller value is used.
   */
  get maxDistance(): number | undefined {
    return this.fields.maxDistance;
  }
  set maxDistance(v: number | undefined) {
    if (!v) {
      delete this.fields.maxDistance;
      return;
    }
    if (!Number.isInteger)
      throw new TypeError(
        'maxDistance must be a positive integer distance in meters!',
      );
    this.fields.maxDistance = v;
  }

  /**
   * Beacons marking locations where the pass is relevant.
   */
  get beacons(): ApplePass['beacons'] {
    return this.fields.beacons;
  }
  set beacons(v: ApplePass['beacons']) {
    if (!v || !Array.isArray(v)) {
      delete this.fields.beacons;
      return;
    }
    for (const beacon of v) {
      if (!beacon.proximityUUID)
        throw new TypeError(`each beacon must contain proximityUUID`);
    }
    // copy array
    this.fields.beacons = [...v];
  }

  /**
   * Information specific to the pass’s barcode.
   * The system uses the first valid barcode dictionary in the array.
   * Additional dictionaries can be added as fallbacks.
   */
  get barcodes(): ApplePass['barcodes'] {
    return this.fields.barcodes;
  }
  set barcodes(v: ApplePass['barcodes']) {
    if (!v) {
      delete this.fields.barcodes;
      delete this.fields.barcode;
      return;
    }

    if (!Array.isArray(v))
      throw new TypeError(`barcodes must be an Array, received ${typeof v}`);

    // Barcodes dictionary: https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3
    for (const barcode of v) {
      if (!BARCODES_FORMAT.has(barcode.format))
        throw new TypeError(
          `Barcode format value ${barcode.format} is invalid!`,
        );
      if (typeof barcode.message !== 'string')
        throw new TypeError('Barcode message string is required');
      if (typeof barcode.messageEncoding !== 'string')
        throw new TypeError('Barcode messageEncoding is required');
    }

    // copy array
    this.fields.barcodes = [...v];
  }

  /**
   * Adds a location where a pass is relevant.
   *
   * @param {number[] | { lat: number, lng: number, alt?: number } | { longitude: number, latitude: number, altitude?: number }} point
   * @param {string} [relevantText]
   * @returns {this}
   */
  addLocation(
    point:
      | number[]
      | { lat: number; lng: number; alt?: number }
      | { longitude: number; latitude: number; altitude?: number },
    relevantText?: string,
  ): this {
    const { longitude, latitude, altitude } = getGeoPoint(point);
    const location: import('../interfaces').Location = {
      longitude,
      latitude,
    };
    if (altitude) location.altitude = altitude;
    if (typeof relevantText === 'string') location.relevantText = relevantText;
    if (!Array.isArray(this.fields.locations))
      this.fields.locations = [location];
    else this.fields.locations.push(location);
    return this;
  }

  get locations(): ApplePass['locations'] {
    return this.fields.locations;
  }

  set locations(v: ApplePass['locations']) {
    delete this.fields.locations;
    if (!v) return;
    if (!Array.isArray(v)) throw new TypeError(`locations must be an array`);
    else for (const location of v) this.addLocation(location);
  }
}
