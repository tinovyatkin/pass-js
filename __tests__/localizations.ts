/* eslint-disable no-shadow */
import * as path from 'path';
import { TextDecoder } from 'util';

import * as Localization from '../src/lib/localizations';

describe('Localizations files helpers', () => {
  it('should read pass.strings file', async () => {
    const resRu = await Localization.readLprojStrings(
      path.resolve(
        __dirname,
        './resources/passes/Generic/ru.lproj/pass.strings',
      ),
    );
    expect([...resRu]).toMatchSnapshot();
    const resZh = await Localization.readLprojStrings(
      path.resolve(
        __dirname,
        './resources/passes/Generic/zh_CN.lproj/pass.strings',
      ),
    );
    expect([...resZh]).toMatchSnapshot();
  });

  it('returns string as utf-16 .lproj buffer', async () => {
    const resRu = await Localization.readLprojStrings(
      path.resolve(
        __dirname,
        './resources/passes/Generic/ru.lproj/pass.strings',
      ),
    );
    const buf = Localization.getLprojBuffer(resRu);
    expect(Buffer.isBuffer(buf)).toBeTruthy();
    const decoder = new TextDecoder('utf-16', { fatal: true });
    expect(decoder.decode(buf)).toMatchSnapshot();
  });

  it('loads localizations from files', async () => {
    const loc = new Localization.Localizations();
    await loc.load(path.resolve(__dirname, './resources/passes/Generic'));
    expect(loc.size).toBe(2);
    // ensure it normalizes locale
    expect(loc.has('zh-CN')).toBeTruthy();
  });
});
