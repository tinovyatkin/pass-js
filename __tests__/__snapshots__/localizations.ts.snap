// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Localizations files helpers returns string as utf-16 .lproj buffer 1`] = `
"\\"Terms of Service\\" = \\"Условия использования\\";
\\"Privacy Policy\\" = \\"Политика конфиденциальности\\";
\\"Legal Notices\\" = \\"Юридические уведомления\\";
\\"Unable to load webpage.\\" = \\"Не удалось загрузить веб-страницу.\\";
\\"OK\\" = \\"ОК\\";
\\"Do you want to add the token named “%@”?\\" = \\"Добавить токен \\"%@\\"?\\";
\\"No\\" = \\"Нет\\";
\\"Yes\\" = \\"Да\\";
\\"Invalid Key\\" = \\"Недействительный ключ\\";
\\"The key '%@' is invalid.\\" = \\"Ключ \\"%@\\" недействителен.\\";
\\"You must enter a key.\\" = \\"Необходимо ввести ключ.\\";
\\"Try Again\\" = \\"Повторить попытку\\";
\\"Cancel\\" = \\"Отмена\\";
\\"Invalid Barcode\\" = \\"Недопустимый штрихкод\\";
\\"The barcode '%@' is not a valid authentication token barcode.\\" = \\"Штрихкод \\"%@\\" не является действительным токеном аутентификации.\\";
\\"Welcome\\" = \\"Добро пожаловать\\";
\\"Welcome_label\\" = \\"Google Authenticator обеспечивает дополнительную безопасность аккаунта с помощью сгенерированных кодов.\\\\n\\\\nЧтобы использовать Google Authenticator, необходимо добавить аккаунт, нажав:\\";
\\"Legal Information\\" = \\"Юридическая информация\\";
\\"Authenticator\\" = \\"Authenticator\\";
\\"Google Authenticator\\" = \\"Google Authenticator\\";
\\"Add Token\\" = \\"Добавить токен\\";
\\"Time Based\\" = \\"По времени\\";
\\"Counter Based\\" = \\"По счетчику\\";
\\"Account:\\" = \\"Аккаунт:\\";
\\"user@example.com\\" = \\"user@example.com\\";
\\"Key:\\" = \\"Ключ:\\";
\\"Enter your key\\" = \\"Введите ключ\\";
\\"Scan Barcode\\" = \\"Сканировать штрихкод\\";
\\"Integrity Check Value\\" = \\"Integrity Check Value\\";"
`;

exports[`Localizations files helpers should read pass.strings file 1`] = `
Array [
  Array [
    "Terms of Service",
    "Условия использования",
  ],
  Array [
    "Privacy Policy",
    "Политика конфиденциальности",
  ],
  Array [
    "Legal Notices",
    "Юридические уведомления",
  ],
  Array [
    "Unable to load webpage.",
    "Не удалось загрузить веб-страницу.",
  ],
  Array [
    "OK",
    "ОК",
  ],
  Array [
    "Do you want to add the token named “%@”?",
    "Добавить токен \\"%@\\"?",
  ],
  Array [
    "No",
    "Нет",
  ],
  Array [
    "Yes",
    "Да",
  ],
  Array [
    "Invalid Key",
    "Недействительный ключ",
  ],
  Array [
    "The key '%@' is invalid.",
    "Ключ \\"%@\\" недействителен.",
  ],
  Array [
    "You must enter a key.",
    "Необходимо ввести ключ.",
  ],
  Array [
    "Try Again",
    "Повторить попытку",
  ],
  Array [
    "Cancel",
    "Отмена",
  ],
  Array [
    "Invalid Barcode",
    "Недопустимый штрихкод",
  ],
  Array [
    "The barcode '%@' is not a valid authentication token barcode.",
    "Штрихкод \\"%@\\" не является действительным токеном аутентификации.",
  ],
  Array [
    "Welcome",
    "Добро пожаловать",
  ],
  Array [
    "Welcome_label",
    "Google Authenticator обеспечивает дополнительную безопасность аккаунта с помощью сгенерированных кодов.

Чтобы использовать Google Authenticator, необходимо добавить аккаунт, нажав:",
  ],
  Array [
    "Legal Information",
    "Юридическая информация",
  ],
  Array [
    "Authenticator",
    "Authenticator",
  ],
  Array [
    "Google Authenticator",
    "Google Authenticator",
  ],
  Array [
    "Add Token",
    "Добавить токен",
  ],
  Array [
    "Time Based",
    "По времени",
  ],
  Array [
    "Counter Based",
    "По счетчику",
  ],
  Array [
    "Account:",
    "Аккаунт:",
  ],
  Array [
    "user@example.com",
    "user@example.com",
  ],
  Array [
    "Key:",
    "Ключ:",
  ],
  Array [
    "Enter your key",
    "Введите ключ",
  ],
  Array [
    "Scan Barcode",
    "Сканировать штрихкод",
  ],
  Array [
    "Integrity Check Value",
    "Integrity Check Value",
  ],
]
`;

exports[`Localizations files helpers should read pass.strings file 2`] = `
Array [
  Array [
    "Terms of Service",
    "服务条款",
  ],
  Array [
    "Privacy Policy",
    "隐私权政策",
  ],
  Array [
    "Legal Notices",
    "法律声明",
  ],
  Array [
    "Unable to load webpage.",
    "无法加载网页。",
  ],
  Array [
    "OK",
    "确定",
  ],
  Array [
    "Do you want to add the token named “%@”?",
    "您想添加名为“%@”的令牌吗？",
  ],
  Array [
    "No",
    "否",
  ],
  Array [
    "Yes",
    "是",
  ],
  Array [
    "Invalid Key",
    "密钥无效",
  ],
  Array [
    "The key '%@' is invalid.",
    "密钥“%@”无效。",
  ],
  Array [
    "You must enter a key.",
    "您必须输入一个密钥。",
  ],
  Array [
    "Try Again",
    "重试",
  ],
  Array [
    "Cancel",
    "取消",
  ],
  Array [
    "Invalid Barcode",
    "条形码无效",
  ],
  Array [
    "The barcode '%@' is not a valid authentication token barcode.",
    "条形码“%@”不是有效的身份验证令牌条形码。",
  ],
  Array [
    "Welcome",
    "欢迎",
  ],
  Array [
    "Welcome_label",
    "Google 身份验证器可以生成验证码，从而增强登录时的安全性。

要使用 Google 身份验证器，必须先添加一个帐户，方法是按：",
  ],
  Array [
    "Legal Information",
    "法律信息",
  ],
  Array [
    "Authenticator",
    "身份验证器",
  ],
  Array [
    "Google Authenticator",
    "Google 身份验证器",
  ],
  Array [
    "Add Token",
    "添加令牌",
  ],
  Array [
    "Time Based",
    "基于时间",
  ],
  Array [
    "Counter Based",
    "基于计数器",
  ],
  Array [
    "Account:",
    "帐户：",
  ],
  Array [
    "user@example.com",
    "user@example.com",
  ],
  Array [
    "Key:",
    "密钥：",
  ],
  Array [
    "Enter your key",
    "输入您的密钥",
  ],
  Array [
    "Scan Barcode",
    "扫描条形码",
  ],
  Array [
    "Integrity Check Value",
    "Integrity Check Value",
  ],
]
`;
